Add docs for writing test cases (#791)
connectrpc/conformance: docs/authoring_test_cases.md

The below `reliesOn` directives are used to signal to the test runner how the reference client or server should be
configured when running tests:

* `reliesOnTls` specifies that a suite relies on TLS. Defaults to `false`.

* `reliesOnTlsClientCerts` specifies that the suite relies on the _client_ using TLS certificates to authenticate with 
  the server. Note that if this is set to `true`, `reliesOnTls` must also be `true`. Defaults to `false`.

* `reliesOnConnectGet` specifies that the suite relies on the Connect GET protocol. Defaults to `false`.

* `reliesOnMessageReceiveLimit` specifies that the suite relies on support for limiting the size of received messages.
  When `true`, the `mode` property must be set to indicate whether the client or server should support the limit. Defaults
  to `false`.

## Test cases

Test cases are specified in the `testCases` property of the suite. Each test case starts with the `request` property 
which defines the request which will be sent to a client during the test run. Each `request` must specify the following
fields:

* `testName` - For naming conventions, see [below](#naming-conventions).

> * `service` - The fully-qualified path of the service this test will interact with, typically the `ConformanceService`.

Reviewer: ```suggestion
* `service` - The fully-qualified name of the service this test will interact with. For now, this is always `connectrpc.conformance.v1.ConformanceService`.
```

* `method` - This is a string specifying the method on `service` that will be called.
* `streamType` - One of `STREAM_TYPE_UNARY`, `STREAM_TYPE_CLIENT_STREAM`, `STREAM_TYPE_SERVER_STREAM`, 
  `STREAM_TYPE_HALF_DUPLEX_BIDI_STREAM`, or `STREAM_TYPE_FULL_DUPLEX_BIDI_STREAM`.

Once the above are specified, you can then define your request. For a full list of fields to specify in the request,

> see the [`ClientCompatRequest`][client-compat-proto] message in the Conformance Protobuf definitions.

Reviewer: I think it will be more helpful to link to the actual message description in the generated docs instead of linking to the source of the containing file
https://buf.build/connectrpc/conformance/docs/main:connectrpc.conformance.v1#connectrpc.conformance.v1.ClientCompatRequest

 > [!IMPORTANT]  
 > The `ClientCompatRequest` message contains some fields that should _not_ be specified in test cases.
 > These fields include:
 > * Fields 1 through 8 in the message definition. These fields are automatically populated by the test runner.
 >   If a test is specific to one of these values, it should instead be indicated in the directives for the test suite
 >   itself.

>  > * Field 20 (`raw_request`). This field is only used by the reference client for sending anomalous requests.

Reviewer: This doesn't really belong in this list. Users will need to specify this in test cases that need to test exceptional wire-format cases.

In fact, it's probably worth a section in here about the use of raw requests (which are only handled by the reference client and should only appear in files with "server" mode) and raw responses (vice versa, only for files with "client" mode, handled by the reference server).
